sql/src/main/java/io/crate/operation/scalar/arithmetic/ArithmeticFunctions.java

```diff
@@ -0,0 +1,177 @@
+
+    final static class DoubleFunctionResolver extends BaseFunctionResolver {
+
+        private static final Signature.ArgMatcher ARITHMETIC_TYPE = Signature.ArgMatcher.of(
+            DataTypes.NUMERIC_PRIMITIVE_TYPES::contains, DataTypes.TIMESTAMP::equals);
+        private final String name;
+        private final BinaryOperator<Double> doubleFunction;
+
+        DoubleFunctionResolver(String name, BinaryOperator<Double> doubleFunction) {
+            super(Signature.of(ARITHMETIC_TYPE, ARITHMETIC_TYPE));
+            this.name = name;
+            this.doubleFunction = doubleFunction;
+        }
+
+        @Override
+        public FunctionImplementation getForTypes(List<DataType> dataTypes) throws IllegalArgumentException {
+            return new BinaryScalar<>(doubleFunction, name, DataTypes.DOUBLE, FunctionInfo.DETERMINISTIC_ONLY);
+        }
+    }
+
+
+    final static class ArithmeticFunctionResolver extends BaseFunctionResolver {
+
+        private static final Signature.ArgMatcher ARITHMETIC_TYPE = Signature.ArgMatcher.of(
+            DataTypes.NUMERIC_PRIMITIVE_TYPES::contains, DataTypes.TIMESTAMP::equals);
+        private final String name;
+        private final String operator;
+        private final Set<FunctionInfo.Feature> features;
+
+        private final BinaryOperator<Double> doubleFunction;
+        private final BinaryOperator<Long> longFunction;
+        private final BinaryOperator<Float> floatFunction;
+
+        ArithmeticFunctionResolver(String name,
+                                   String operator,
+                                   Set<FunctionInfo.Feature> features,
+                                   BinaryOperator<Double> doubleFunction,
+                                   BinaryOperator<Long> longFunction,
+                                   BinaryOperator<Float> floatFunction) {
+            super(Signature.of(ARITHMETIC_TYPE, ARITHMETIC_TYPE));
+            this.name = name;
+            this.operator = operator;
+            this.doubleFunction = doubleFunction;
+            this.longFunction = longFunction;
+            this.floatFunction = floatFunction;
+            this.features = features;
+        }
+
+        @Override
+        public FunctionImplementation getForTypes(List<DataType> dataTypes) throws IllegalArgumentException {
+            BinaryScalar<?> scalar;
+            if (containsTypesWithDecimal(dataTypes)) {
+                if (containsDouble(dataTypes)) {
+                    scalar = new BinaryScalar<>(doubleFunction, name, DataTypes.DOUBLE, features);
+                } else {
+                    scalar = new BinaryScalar<>(floatFunction, name, DataTypes.FLOAT, features);
+                }
+            } else {
+                scalar = new BinaryScalar<>(longFunction, name, DataTypes.LONG, features);
+            }
+            return Scalar.withOperator(scalar, operator);
+        }
+
+    }
+
+    public static Function of(String name, Symbol first, Symbol second) {
```

Comment:
the functionInfo features used are Add specific - maybe rename the method to `forAdd` and remove the name argument? If this is required for other functions, maybe have the features as argument?